File: app/src/common/shared/com/igalia/wolvic/search/SearchEngineWrapper.java
package com.igalia.wolvic.search;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.util.Log;

import androidx.annotation.NonNull;

import com.igalia.wolvic.R;
import com.igalia.wolvic.VRBrowserActivity;
import com.igalia.wolvic.browser.SettingsStore;
import com.igalia.wolvic.geolocation.GeolocationData;
import com.igalia.wolvic.search.suggestions.SearchSuggestionsClientKt;
import com.igalia.wolvic.utils.SystemUtils;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;

import kotlin.coroutines.CoroutineContext;
import kotlinx.coroutines.Dispatchers;
import mozilla.components.browser.search.SearchEngine;
import mozilla.components.browser.search.SearchEngineManager;
import mozilla.components.browser.search.provider.AssetsSearchEngineProvider;
import mozilla.components.browser.search.provider.filter.SearchEngineFilter;
import mozilla.components.browser.search.provider.localization.LocaleSearchLocalizationProvider;
import mozilla.components.browser.search.provider.localization.SearchLocalizationProvider;
import mozilla.components.browser.search.suggestions.SearchSuggestionClient;

public class SearchEngineWrapper implements SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String LOGTAG = SystemUtils.createLogtag(SearchEngineWrapper.class);

    private static SearchEngineWrapper mSearchEngineWrapperInstance;

    public static synchronized @NonNull
    SearchEngineWrapper get(final @NonNull Context aContext) {
        if (mSearchEngineWrapperInstance == null) {
            mSearchEngineWrapperInstance = new SearchEngineWrapper(aContext);
        } else if (!aContext.equals(mSearchEngineWrapperInstance.getContext())) {
            // Because of the architecture of the app, this is very unlikely (but just in case...).
            Log.w(LOGTAG, "Context has changed");
            mSearchEngineWrapperInstance = new SearchEngineWrapper(aContext);
        }

        return mSearchEngineWrapperInstance;
    }

    private final WeakReference<Context> mContextRef;
    private SearchEngine mSearchEngine;
    private SearchSuggestionClient mSuggestionsClient;
    private final SharedPreferences mPrefs;
    private boolean mAutocompleteEnabled;
    private HashMap<String, SearchEngine> mSearchEnginesMap;

    private SearchEngineWrapper(@NonNull Context aContext) {
        mContextRef = new WeakReference<>(aContext);
        mPrefs = PreferenceManager.getDefaultSharedPreferences(getContext());
        mAutocompleteEnabled = SettingsStore.getInstance(getContext()).isAutocompleteEnabled();

        String preferredSearchEngineId = SettingsStore.getInstance(getContext()).getSearchEngineId();

        setupSearchEngine(aContext, preferredSearchEngineId);
    }

    public void registerForUpdates() {
        if (hasContext()) {
            getContext().registerReceiver(
                    mLocaleChangedReceiver,
                    new IntentFilter(Intent.ACTION_LOCALE_CHANGED));
            if (mPrefs != null) {
                mPrefs.registerOnSharedPreferenceChangeListener(this);
            }
        }
    }

    public void unregisterForUpdates() {
        if (hasContext()) {
            try {
                getContext().unregisterReceiver(mLocaleChangedReceiver);
            } catch (IllegalArgumentException ignored) {
            }
            if (mPrefs != null) {
                mPrefs.unregisterOnSharedPreferenceChangeListener(this);
            }
        }
    }

    public String getSearchURL(String aQuery) {
        return mSearchEngine.buildSearchUrl(aQuery);
    }

    public CompletableFuture<List<String>> getSuggestions(String aQuery) {
        return SearchSuggestionsClientKt.getSuggestionsAsync(mSuggestionsClient, aQuery != null ? aQuery : "");
    }

    public String getResourceURL() {
        Uri uri = Uri.parse(mSearchEngine.buildSearchUrl(""));
        return uri.getScheme() + "://" + uri.getHost();
    }

    public SearchEngine getCurrentSearchEngine() {
        return mSearchEngine;
    }

    // Receiver for locale updates
    private BroadcastReceiver mLocaleChangedReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction().equals(Intent.ACTION_LOCALE_CHANGED)) {
                String userSearchEngineId = SettingsStore.getInstance(getContext()).getSearchEngineId();
                setupSearchEngine(context, userSearchEngineId);
            }
        }
    };

    public Collection<SearchEngine> getAvailableSearchEngines() {
        return mSearchEnginesMap.values();
    }

    public void setDefaultSearchEngine() {
        if (hasContext())
            setupSearchEngine(getContext(), null);
    }

    public void setCurrentSearchEngineId(Context context, String searchEngineId) {
        if (hasContext())
            SettingsStore.getInstance(getContext()).setSearchEngineId(searchEngineId);
    }

    /**
     * We cannot send system ACTION_LOCALE_CHANGED so the component refreshes the engines
     * with the updated SearchLocalizationProvider information so we have to update the whole manager.
     *
     * @param aContext Activity context
     * @param userPref User preferred engine (among the available ones)
     */
    private void setupSearchEngine(@NonNull Context aContext, String userPref) {
        List<SearchEngineFilter> engineFilterList = new ArrayList<>();

        GeolocationData data = GeolocationData.parse(SettingsStore.getInstance(aContext).getGeolocationData());
        SearchLocalizationProvider localizationProvider;
        if (data == null) {
            Log.d(LOGTAG, "Using Locale based search localization provider");
            // If we don't have geolocation data we default to the Locale search localization provider
            localizationProvider = new LocaleSearchLocalizationProvider();
        } else {
            Log.d(LOGTAG, "Using Geolocation based search localization provider: " + data);
            // If we have geolocation data we initialize the provider with the received data.
            localizationProvider = new GeolocationLocalizationProvider(data);
        }

        // Configure the assets search with the localization provider and the engines that we want
        // to filter.
        AssetsSearchEngineProvider engineProvider = new AssetsSearchEngineProvider(
                localizationProvider,
                Collections.emptyList(), //engineFilterList,
                Collections.emptyList());

        SearchEngineManager searchEngineManager = new SearchEngineManager(Collections.singletonList(engineProvider),
                (CoroutineContext) Dispatchers.getDefault());

        // If we don't get any result we use the default configuration.
        List<SearchEngine> searchEngines = searchEngineManager.getSearchEngines(aContext);
        if (searchEngines.size() == 0) {
            Log.d(LOGTAG, "  Could not find any available search engines, using default.");
            searchEngineManager = new SearchEngineManager();
            searchEngines = searchEngineManager.getSearchEngines(aContext);
        }

        mSearchEnginesMap = new LinkedHashMap<>(searchEngines.size());
        for (int i = 0; i < searchEngines.size(); i++) {
            SearchEngine searchEngine = searchEngines.get(i);
            mSearchEnginesMap.put(searchEngine.getIdentifier(), searchEngine);
        }

        String userPrefName = "";
        if (mSearchEnginesMap.containsKey(userPref)) {
            // The search component API uses the engine's name, not its identifier.
            userPrefName = mSearchEnginesMap.get(userPref).getName();
        }

        // A name can be used if the user gets to choose among the available engines
        mSearchEngine = searchEngineManager.getDefaultSearchEngine(aContext, userPrefName);

        mSuggestionsClient = new SearchSuggestionClient(mSearchEngine,
                (searchUrl, continuation) -> {
                    if (mAutocompleteEnabled && getVRBrowserActivity() != null) {
                        if (!getVRBrowserActivity().getWindows().isInPrivateMode()) {
                            return SearchSuggestionsClientKt.fetchSearchSuggestions(getContext(), searchUrl);
                        }
                    }
                    return null;
                }
        );
        SettingsStore.getInstance(getContext()).setSearchEngineId(mSearchEngine.getIdentifier());
    }

    private boolean hasContext() {
        return mContextRef.get() != null;
    }

    private Context getContext() {
        return mContextRef.get();
    }

    private VRBrowserActivity getVRBrowserActivity() {
        if (hasContext() && getContext() instanceof VRBrowserActivity)
            return (VRBrowserActivity) getContext();
        else
            return null;
    }

    // SharedPreferences.OnSharedPreferenceChangeListener

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (mContextRef.get() != null) {
            if (key.equals(mContextRef.get().getString(R.string.settings_key_geolocation_data)) ||
                    key.equals(mContextRef.get().getString(R.string.settings_key_search_engine_id))) {
                String searchEngineId = SettingsStore.getInstance(mContextRef.get()).getSearchEngineId();
                if (mSearchEngine == null || !Objects.equals(mSearchEngine.getIdentifier(), searchEngineId)) {
                    setupSearchEngine(mContextRef.get(), searchEngineId);
                }
            } else if (key.equals(mContextRef.get().getString(R.string.settings_key_autocomplete))) {
                mAutocompleteEnabled = SettingsStore.getInstance(mContextRef.get()).isAutocompleteEnabled();
            }
        }
    }
}

File: app/src/common/shared/com/igalia/wolvic/search/SearchEngineWrapper.kt
package com.igalia.wolvic.search

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.net.Uri
import android.os.Build
import android.util.Log
import androidx.preference.PreferenceManager
import com.igalia.wolvic.R
import com.igalia.wolvic.VRBrowserActivity
import com.igalia.wolvic.browser.SettingsStore
import com.igalia.wolvic.geolocation.GeolocationData
import com.igalia.wolvic.search.suggestions.fetchSearchSuggestions
import com.igalia.wolvic.search.suggestions.getSuggestionsAsync
import com.igalia.wolvic.utils.SystemUtils
import kotlinx.coroutines.Dispatchers
import mozilla.components.browser.state.action.SearchAction
import mozilla.components.browser.state.search.RegionState
import mozilla.components.browser.state.search.SearchEngine
import mozilla.components.browser.state.state.searchEngines
import mozilla.components.browser.state.store.BrowserStore
import mozilla.components.feature.search.ext.buildSearchUrl
import mozilla.components.feature.search.middleware.SearchMiddleware
import mozilla.components.feature.search.suggestions.SearchSuggestionClient
import java.lang.ref.WeakReference
import java.util.concurrent.CompletableFuture
import kotlin.coroutines.CoroutineContext

class SearchEngineWrapper private constructor(aContext: Context) :
    OnSharedPreferenceChangeListener {
    private val mContextRef: WeakReference<Context?>
    private var currentSearchEngine: SearchEngine? = null
    private var mSuggestionsClient: SearchSuggestionClient? = null
    private val mPrefs: SharedPreferences?
    private var mAutocompleteEnabled: Boolean
    private var mSearchEnginesMap: LinkedHashMap<String?, SearchEngine>? = null
    private val mIoDispatcher: CoroutineContext = Dispatchers.IO
    private var mRegionState: RegionState? = null
    private val mBrowserStore = BrowserStore(
        middleware = listOf(
            SearchMiddleware(aContext,
                ioDispatcher = mIoDispatcher)
        ))
    fun registerForUpdates() {
        if (hasContext()) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                context!!.registerReceiver(
                    mLocaleChangedReceiver,
                    IntentFilter(Intent.ACTION_LOCALE_CHANGED), Context.RECEIVER_NOT_EXPORTED
                )
            } else {
                context!!.registerReceiver(
                        mLocaleChangedReceiver,
                        IntentFilter(Intent.ACTION_LOCALE_CHANGED)
                )
            }
            mPrefs?.registerOnSharedPreferenceChangeListener(this)
        }
    }

    fun setCurrentSearchEngine(searchEngine: SearchEngine?) {
        currentSearchEngine = searchEngine
    }

    fun unregisterForUpdates() {
        if (hasContext()) {
            try {
                context!!.unregisterReceiver(mLocaleChangedReceiver)
            } catch (ignored: IllegalArgumentException) {
            }
            mPrefs?.unregisterOnSharedPreferenceChangeListener(this)
        }
    }

    fun getSearchURL(aQuery: String?): String {
        if (currentSearchEngine == null) setupPreferredSearchEngine()
        return currentSearchEngine!!.buildSearchUrl(aQuery!!)
    }

    fun getSuggestions(aQuery: String?): CompletableFuture<List<String>?> {
        return getSuggestionsAsync(mSuggestionsClient!!, aQuery ?: "")
    }

    val resourceURL: String
        get() {
            if (currentSearchEngine == null) setupPreferredSearchEngine()
            val uri = Uri.parse(currentSearchEngine!!.buildSearchUrl(""))
            return uri.scheme + "://" + uri.host
        }

    // Receiver for locale updates
    private val mLocaleChangedReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == Intent.ACTION_LOCALE_CHANGED) {
                setupPreferredSearchEngine()
            }
        }
    }

    init {
        mContextRef = WeakReference(aContext)
        mPrefs = PreferenceManager.getDefaultSharedPreferences(context!!)
        mAutocompleteEnabled = SettingsStore.getInstance(context!!).isAutocompleteEnabled
        setupPreferredSearchEngine()
    }

    val availableSearchEngines: Collection<SearchEngine>
        get() {
            updateSearchEngine()
            return mSearchEnginesMap!!.values
        }

    fun setDefaultSearchEngine() {
        if (hasContext()) setupSearchEngine(context!!, null)
    }

    fun setCurrentSearchEngineId(context: Context?, searchEngineId: String?) {
        if (hasContext()) SettingsStore.getInstance(context!!).searchEngineId = searchEngineId
    }

    fun setupPreferredSearchEngine() {
        val preferredSearchEngineId = SettingsStore.getInstance(context!!).searchEngineId
        setupSearchEngine(context!!, preferredSearchEngineId)
    }

    /**
     * We cannot directly getCurrentSearchEngine() if it's not fully initialized, so use this method instead.
     *
     * @return SearchEngine currentSearchEngine
     */
    fun resolveCurrentSearchEngine(): SearchEngine {
        if (currentSearchEngine == null) setupPreferredSearchEngine()
        return currentSearchEngine!!
    }

    /**
     * We cannot send system ACTION_LOCALE_CHANGED when the component refreshes the engines
     * with the updated information so we have to update the whole manager.
     *
     * @param aContext Activity context
     * @param userPref User preferred engine (among the available ones)
     */
    private fun setupSearchEngine(aContext: Context?, userPref: String?) {
        val thisContext: Context = aContext ?: context!!

        val data = GeolocationData.parse(SettingsStore.getInstance(thisContext).geolocationData)
        val regionState = if (data != null) {
            Log.d(LOGTAG, "Using Geolocation based search localization: $data")
            // If we have geolocation data we initialize with the received data.
            val country = data.countryCode
            RegionState(country, country)
        } else {
            val country = thisContext.resources.configuration.locales.get(0).country
            RegionState(country, country)
        }

        if (mRegionState == null || mRegionState != regionState) {
            mBrowserStore.dispatch(
                SearchAction.SetRegionAction(
                    regionState,
                ),
            )
            mRegionState = regionState
        }

        val searchEngines: List<SearchEngine> = mBrowserStore.state.search.searchEngines
        mSearchEnginesMap = LinkedHashMap(searchEngines.size)
        for (i in searchEngines.indices) {
            val searchEngine = searchEngines[i]
            mSearchEnginesMap!![searchEngine.id] = searchEngine
        }

        val newSearchEngine = if (mSearchEnginesMap!!.containsKey(userPref)) {
            mSearchEnginesMap!![userPref]
        } else {
            mSearchEnginesMap!![mBrowserStore.state.search.regionDefaultSearchEngineId]
        }

        if (newSearchEngine == null || currentSearchEngine == newSearchEngine) return

        mSuggestionsClient = SearchSuggestionClient(
            newSearchEngine
        ) label@
        { searchUrl: String? ->
            if (mAutocompleteEnabled && vRBrowserActivity != null) {
                if (!vRBrowserActivity!!.windows.isInPrivateMode) {
                    return@label fetchSearchSuggestions(thisContext, searchUrl!!)
                }
            }
            null
        }
        currentSearchEngine = newSearchEngine
    }

    private fun hasContext(): Boolean {
        return mContextRef.get() != null
    }

    private fun updateSearchEngine() {
        val newSearchEngineName = if (currentSearchEngine != null) {
            currentSearchEngine!!.id
        } else {
            null
        }

        setupSearchEngine(null, newSearchEngineName)
    }

    private val context: Context?
        private get() = mContextRef.get()
    private val vRBrowserActivity: VRBrowserActivity?
        private get() = if (hasContext() && context is VRBrowserActivity) context as VRBrowserActivity? else null

    // SharedPreferences.OnSharedPreferenceChangeListener
    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        if (mContextRef.get() != null) {
            if (key == mContextRef.get()!!
                    .getString(R.string.settings_key_geolocation_data) || key == mContextRef.get()!!
                    .getString(R.string.settings_key_search_engine_id)
            ) {
                val searchEngineId = SettingsStore.getInstance(mContextRef.get()!!).searchEngineId
                if (currentSearchEngine == null || currentSearchEngine!!.id != searchEngineId) {
                    setupSearchEngine(mContextRef.get()!!, searchEngineId)
                }
            } else if (key == mContextRef.get()!!.getString(R.string.settings_key_autocomplete)) {
                mAutocompleteEnabled =
                    SettingsStore.getInstance(mContextRef.get()!!).isAutocompleteEnabled
            }
        }
    }

    companion object {
        private val LOGTAG = SystemUtils.createLogtag(SearchEngineWrapper::class.java)
        private var mSearchEngineWrapperInstance: SearchEngineWrapper? = null

        @JvmStatic
        @Synchronized
        operator fun get(aContext: Context): SearchEngineWrapper {
            if (mSearchEngineWrapperInstance == null) {
                mSearchEngineWrapperInstance = SearchEngineWrapper(aContext)
            } else if (aContext != mSearchEngineWrapperInstance!!.context) {
                // Because of the architecture of the app, this is very unlikely (but just in case...).
                Log.w(LOGTAG, "Context has changed")
                mSearchEngineWrapperInstance = SearchEngineWrapper(aContext)
            }
            return mSearchEngineWrapperInstance!!
        }
    }
}
